Proposal: Arm64 Support and Multi-Architecture Image Publishing for Harbor

ranimandepudi opened the pull request:
Tested and confirmed: Harbor functions fully on both Arm64 and Amd64 platforms, including the UI, API, registry, and multi-architecture artifact handling.

As the author of this contribution from Arm, I welcome any feedback, collaboration, or questions from the Harbor community.

Thank you to the community for your continued support. Arm ambassadors are always ready to support Harbor. 


Related PR in the main Harbor repository: [goharbor/harbor#21982](https://github.com/goharbor/harbor/pull/21982)

ranimandepudi commented:
> @ranimandepudi bitnami has already built arm images for Harbor, it is open-sourced in https://github.com/bitnami/containers/tree/main/bitnami, and images is available in https://hub.docker.com/r/bitnami/harbor-core, maybe you could reference it.

Thanks for sharing this @stonezdj !

Just to clarify — the changes I proposed in the PR are focused solely on enabling multi-arch Docker image builds without impacting existing functionality. Since Photon OS already supports both amd64 and arm64, my patch preserves the current behavior while adding support for Arm builds in a clean and upstream-friendly way.

OrlinVasilev commented:
@ranimandepudi can you link the PR in the main repository for the changes!
I find it super useful if we as a project can provide our arm64 arch images! +1

OrlinVasilev commented:
@ranimandepudi can you please fix your DCO

OrlinVasilev reviewed (approved):
I think it's a MUST have!

ranimandepudi commented:
@OrlinVasilev Thank you for your support! I have made the updates based on your comments. Please let me know if any further changes are needed.

wy65701436 commented on proposals/arm64-support-harbor.md:
```diff
@@ -0,0 +1,137 @@
+
+## Design Goals
+
+Build Harbor component images (core, jobservice, registryctl, etc.) for both Arm64 and Amd64 using native compilation per architecture and multi-arch manifest merging.
```
Currently, each pull request triggers multiple pipelines to test the code. We need to ensure that at least one of these pipelines runs on an ARM-based machine to validate code changes. Including this in the proposal should be a mandatory requirement.

wy65701436 commented on proposals/arm64-support-harbor.md:
```diff
@@ -0,0 +1,137 @@
+-  Build image: docker build -t ranichowdary/core-harbor:arm64 .
+- Push the docker push ranichowdary/core-harbor:arm64
+
+Created manifest to ensure final image : `latest` tag resolves to the correct architecture during image pull based on the target environment.
```
Do you mean that we ultimately use an image index with the `latest` tag to include both the x64 and arm64 image references?

ranimandepudi replied:
Yes, exactly — the goal is to publish a multi-architecture image index (a Docker manifest list) under the same latest tag, which references both the amd64 and arm64 images.

You're right that we haven’t yet explicitly proposed using docker manifest create, but that would be the correct approach to combine the architecture-specific images into a single tag. We can incorporate that into the release process.
Example
```
docker manifest create harbor:latest \
  --amend harbor:latest-amd64 \
  --amend harbor:latest-arm64

docker manifest push harbor:latest
```

wy65701436 commented on proposals/arm64-support-harbor.md:
```diff
@@ -0,0 +1,137 @@
+
+## Testing 
+
+To validate that our custom-built Harbor instance with ARM64 support works correctly across components and architectures, I performed the following levels of testing:
```
This should be automatically done in the CI.

ranimandepudi replied:
Yes I will update it but just added it is also manually tested !!

wy65701436 commented on proposals/arm64-support-harbor.md:
```diff
@@ -0,0 +1,137 @@
+ranichowdary/trivy-base-harbor
+
+
+## Build and Push Workflow (Per Architecture)
```
General comment: This section is a key part of the proposal as it should clearly demonstrate how the solution will be implemented. However, it currently lacks sufficient detail. Would you mind adding more specifics? I think it would be helpful to walk through one image as an example to clarify the process.

wy65701436 commented on proposals/arm64-support-harbor.md:
```diff
@@ -0,0 +1,137 @@
+
+To validate that our custom-built Harbor instance with ARM64 support works correctly across components and architectures, I performed the following levels of testing:
+
+Container Health Validation:
```
Since we aim to enable the entire build process through Make, please ensure that all steps required to build the images from scratch are covered. In addition to the source code, please also consider tools that may be architecture-sensitive—for example, https://github.com/goharbor/harbor/blob/main/tools/spectral/Dockerfile.

ranimandepudi replied:
Thanks for the note — that’s a great point.

My plan is to ensure that all Dockerfiles and supporting tools (including those under tools/) are validated for multi-architecture compatibility. I'll audit components like tools/spectral/Dockerfile and any other architecture-sensitive containers, and make the necessary changes to support arm64 where needed. I will update that in the commit.

wy65701436 replied:
Mark this as unresolved.

Let me give an example: in a previous release, we gracefully added support for ARM64—for instance, by introducing architecture-specific logic in the Dockerfile.

https://github.com/goharbor/harbor/blob/main/tools/spectral/Dockerfile

What I’m looking for now are the technical details. Similar to how we handled the Dockerfile, how should we manage the source code, tool binaries, and image building—end to end? And how should we trigger the full flow from the make command?

wy65701436 replied:
@ranimandepudi can you add the chart into the proposal?

ranimandepudi replied:
@wy65701436  I have updated it !!

wy65701436 commented on proposals/arm64-support-harbor.md:
```diff
@@ -0,0 +1,137 @@
+-  Build image: docker build -t ranichowdary/core-harbor:arm64 .
+- Push the docker push ranichowdary/core-harbor:arm64
+
+Created manifest to ensure final image : `latest` tag resolves to the correct architecture during image pull based on the target environment.
```
I assume we will also display the `latest` tag, similar to how hello-world does.

![image](https://github.com/user-attachments/assets/ce9fada9-b266-4c66-bcf9-f706f07b2ae7)

ranimandepudi replied:
Yes, exactly — our goal is to have the latest tag function the same way. Once we generate and push a multi-architecture manifest, harbor:latest will behave like hello-world:latest, automatically resolving to the appropriate architecture based on the user's platform.

wy65701436 commented on proposals/arm64-support-harbor.md:
```diff
@@ -0,0 +1,137 @@
+ranichowdary/jobservice-harbor
+ranichowdary/jobservice-base-harbor
+Registry:
+ranichowdary/registry-harbor
```
Let me use the registry build as an example to illustrate the idea. We currently have two Dockerfiles involved in the process—one for building the binary and another for creating the final image.

From my understanding, the first step should be updating the Dockerfile.binary and its corresponding builder scripts to support specifying GOARCH when compiling the registry binary. Once that's in place, we can evaluate whether changes to the second Dockerfile (used for building the final image) are also needed.

`RUN CGO_ENABLED=0 GOARCH=$TARGETARCH make PREFIX=/go clean binaries`

I believe it's important to include sufficient technical detail here, rather than keeping it high-level. This will help the maintainers better assess whether the proposal is technically sound and practically feasible.

ranimandepudi replied:
Yes, you're right — the GOARCH=$TARGETARCH assignment is currently missing. This is essential for ensuring the binary is correctly compiled for the target architecture during multi-platform builds.

I will update it and commit !!

ranimandepudi replied:
@wy65701436  While our CI now builds on native amd64 and arm64 runners via a matrix strategy (so cross-compilation isn't strictly required), would you still recommend explicitly adding GOARCH=$TARGETARCH in Dockerfile.binary?

wy65701436 replied:
I think we should have some discussion around this part, as the parameter needs to be passed down from the top-level make call to the lower levels.

A few questions to consider:
    How should we structure the code?
    Will we maintain separate builders and Dockerfiles for each architecture, or consolidate them into one?

From an engineering perspective, I’d prefer to separate the code paths initially — one for x86 and one for ARM — and merge them later once the ARM path is stable and production-ready. While this approach introduces some code duplication, it ensures that our current stable releases remain unaffected.

wy65701436 replied:
@ranimandepudi update on this part?

wy65701436 commented on proposals/arm64-support-harbor.md:
```diff
@@ -0,0 +1,137 @@
+
+## Build and Push Workflow (Per Architecture)
+
+Each image is built and pushed in the following steps:
```
From the entrypoint perspective, will we add a flag to the Makefile to specify the target platform architecture for building images? I assume we should include such a flag, with x86_64 as the default, and allow it to be extended to support arm as well.

wy65701436 replied:
@ranimandepudi any updates?

ranimandepudi replied:
https://github.com/goharbor/harbor/pull/21982/files

Refer to this makefile: I have already added up `ARCH ?= $(shell uname -m)` which auto detects the current machine's arch using uname -m and assigns it to the ARCH variable !! Let me know if you would prefer renaming it to TARGETARCH for consistency with the Docker's conventions

ranimandepudi commented:
@wy65701436 — Could you please review it and let me know if anything needs to be updated ?
Also, you can check this PR: https://github.com/goharbor/harbor/pull/21982/files
which doesn’t have a CI pipeline. However, if you pull the PR and test it — since there are no functional changes — it should work fine on both Arm and Amd instances.

I have the CI pipeline changes in my local, can push them once this is tested !!

MinerYang commented:
Hi @ranimandepudi ,

By reviewing this proposal and the [PR](https://github.com/goharbor/harbor/pull/21982). Besides what we have discussed above(though some of them are still unresolved), I have a few other details that  might need mentioned here
- I believe we should identify all the binray that needs by harbor and its source during the building process. e.g. While we build trivy-adapter, it actually needs 2 binaries, one is Trivy itself, which we always download from upstream, another is for trivy-adpater, it could choose to be built from src or downloading from upstream `https://github.com/goharbor/harbor-scanner-trivy`,  In this case, we should consider to have a arm64 version of it as well.
- There are few comments added in old PR https://github.com/goharbor/harbor/pull/21825  by @bitsf  which still have the same problems in the new one. As you were saying in the proposal it would not block the current build, so basically we don't want change any image name, build tag and all other staff if not necessary.
- From a overall perspective,  I prefer we have a structure design to refine our whole build process.
i. first having an agreement on how should all the go build binaries adopt in arm64 and amd64, e.g.should the `GOARCH` and the `GOOS` been passed explicitly? 
ii. then how do we refine and build all the Dockerfile.base, Dockerfile.
iii. How does the git action and our nightly e2e well tested the harbor instance deployed by these multi-arch images. There's might be more subtle issue will occur like python or Node.js native extensions may fail when running more specific robot cases. We need more investigation and robust test structure to support it.

sudiptosarkar commented:
I was wondering why it was crashlooping on my Raspberry PI k8s cluster. And then I found this. We absolutely do need this...

ranimandepudi commented:
@wy65701436 
 I have addressed all the comments and let me know if there are any. 
 Also addressing the codebase per arch:
 
 Just to clarify - since we are targeting multi-arch support, we should ideally keep a single shared codebase and only add arch-specific logic when truly needed (eg: base image, platform-specific binaries). Were you referring to how we organize the build scripts or Dockerfiles to support both archs cleanly?
 Docker images are multi-arch so when its pulled it pull the arch specific one out of 2 !!

ranimandepudi commented on proposals/arm64-support-harbor.md:
```diff
@@ -0,0 +1,137 @@
+
+These changes allow Harbor to be built, packaged, and installed cleanly on Arm64 hosts such as AWS Graviton and also on x86 as images are multi arch. 
+
+This proposal introduces native Arm64 builds for Harbor components using `Docker Build` and multi-architecture manifests. Each image is tested and published in a way that maintains compatibility with existing Amd64-based deployments.
```
@wy65701436 I already mentioned multiple times that functionally the changes don't alter the Harbor's core behavior - the same code works for both offline and online installers(I have manually tested it) with multi-arch images.  `prepare` is also included in multi-arch build process so offline installers also working on Arm.

wy65701436 commented:
> @wy65701436 The current plan is to release this code under a separate branch initially. This will allow developers and the community to test it thoroughly without impacting the main branch. Once it's been validated and proven stable, we can then integrate into the mainline codebase. This approach ensures stability while still moving forward with multi-arch support.

I understand that we could use a separate code branch for validation. However, from a feature perspective, I would prefer to see a complete technical plan with concrete details. If we proceed with a separate branch, what is the long-term plan? Are we aiming to eventually ship it? Given that we’ve never shipped anything outside of a release branch before, how should this be handled within our current release process?

cc @Vad1mo any idea?

wy65701436 commented:
@ranimandepudi One general question: how do users consume these Arm images? In other words, what kind of installer do we plan to support for Harbor on Arm—Online, Offline, or Helm Chart?

wy65701436 commented:
> @wy65701436 Just to clarify In the current Phase 1 - only focused on enabling multi-arch image builds through: Adding ARCH support and setting up docker images No functionality changes done.
> 
> In the next phase - will introduce the CI pipeline changes to fully integrate multi-arch builds in the CI process. This phased approach ensures stability and avoids impacting current builds !!

can you add this into proposal? How many phases we have, what we do in the phase1?

chlins commented:
Is it possible to further detail the current flowchart by specifying branches for different conditions and the final actions to be executed? Additionally, could you summarize the points that need to be changed at present, and clarify whether there are any outstanding items in the current plan that will need to be addressed in phase 2?

wy65701436 commented:
> Is it possible to further detail the current flowchart by specifying branches for different conditions and the final actions to be executed? Additionally, could you summarize the points that need to be changed at present, and clarify whether there are any outstanding items in the current plan that will need to be addressed in phase 2?



> Hi @ranimandepudi ,
> 
> By reviewing this proposal and the [PR](https://github.com/goharbor/harbor/pull/21982). Besides what we have discussed above(though some of them are still unresolved), I have a few other details that might need mentioned here
> 
>     * I believe we should identify all the binray that needs by harbor and its source during the building process. e.g. While we build trivy-adapter, it actually needs 2 binaries, one is Trivy itself, which we always download from upstream, another is for trivy-adpater, it could choose to be built from src or downloading from upstream `https://github.com/goharbor/harbor-scanner-trivy`,  In this case, we should consider to have a arm64 version of it as well.
> 
>     * There are few comments added in old PR [Harbor on Multi-arch (arm64 & amd64) harbor#21825](https://github.com/goharbor/harbor/pull/21825)  by @bitsf  which still have the same problems in the new one. As you were saying in the proposal it would not block the current build, so basically we don't want change any image name, build tag and all other staff if not necessary.
> 
>     * From a overall perspective,  I prefer we have a structure design to refine our whole build process.
>       i. first having an agreement on how should all the go build binaries adopt for arm64 and amd64 build process, e.g.should the `GOARCH` and the `GOOS` been passed explicitly?
>       ii. then how do we refine and build all the Dockerfile.base, Dockerfile etc..
>       iii. How does the git action and our nightly e2e well tested the harbor instance deployed by these multi-arch images. There's might be more subtle issue will occur like python or Node.js native extensions may fail when running more specific robot cases. We need thorough  investigation and a robust test structure to support it.

@ranimandepudi Points i, ii, and iii are great questions and highlight key aspects we’d like to see addressed in the proposal. Could you please incorporate your answers to these into the proposal?

ranimandepudi commented:
@wy65701436 
I can add them to the proposal and I already discussed them in the comments :
Firstly I will discuss each step detailed here:

- I believe we should identify all the binray that needs by harbor and its source during the building process. e.g. While we build trivy-adapter, it actually needs 2 binaries, one is Trivy itself, which we always download from upstream, another is for trivy-adpater, it could choose to be built from src or downloading from upstream https://github.com/goharbor/harbor-scanner-trivy, In this case, we should consider to have a arm64 version of it as well.

I agree with it also I have build multi-arch images to all of them. These include
All images including binaries, all we have is:
goharbor/prepare - docker pull ranichowdary/prepare-harbor:latest
goharbor/harbor-portal - docker pull ranichowdary/portal-harbor:latest
goharbor/harbor-core - docker pull ranichowdary/core-harbor:latest
goharbor/harbor-jobservice - docker pull ranichowdary/jobservice-harbor:latest
goharbor/harbor-log -docker pull ranichowdary/log-harbor:latest
goharbor/harbor-db - docker pull ranichowdary/db-harbor:latest
goharbor/postgresql-photon -docker pull ranichowdary/db-harbor:latest
goharbor/trivy-adapter-photon -docker pull ranichowdary/trivy-adapter-harbor:latest
goharbor/nginx-photon
goharbor/registry-photon
goharbor/harbor-registryctl
goharbor/redis-photon
goharbor/standalone-db-migrator
goharbor/harbor-exporter
goharbor/harbor-core-base
goharbor/harbor-jobservice-base
goharbor/harbor-registry-base
goharbor/harbor-registryctl-base
goharbor/harbor-db-base
goharbor/harbor-trivy-adapter-base - docker pull ranichowdary/trivy-base-harbor:latest
goharbor/harbor-redis-base
goharbor/harbor-nginx-base
goharbor/harbor-exporter-base
Photon:5.0
Golang:1.23.8
Node:16.18.0
Rest of them also present in my docker-hub - https://hub.docker.com/repositories/ranichowdary

- There are few comments added in old PR https://github.com/goharbor/harbor/pull/21825 by @bitsf which still have the same problems in the new one. As you were saying in the proposal it would not block the current build, so basically we don't want change any image name, build tag and all other staff if not necessary.

I am not blocking the current build - Image names are changed as I can push images to goharbor - I pushed them to my dockerhub and used

We can push them as multi-arch images and push to goharbor so image names or other names dont change and remain unchanged.

Each image is built and pushed in the following steps:
On Amd64 Host:
Generated amd64 binary.
Build image: docker build -t ranichowdary/core-harbor:amd64 .
Push: docker push ranichowdary/core-harbor:amd64
On Arm64 Host:
Run make build-core on Arm64 instance to generate native binary.
Build image: docker build -t ranichowdary/core-harbor:arm64 .
Push: docker push ranichowdary/core-harbor:arm64
Created manifest to ensure final image : latest tag resolves to the correct architecture during image pull based on the target environment.

- From a overall perspective, I prefer we have a structure design to refine our whole build process.

GOARCH/GOOS handling: Since builds were done on native Arm64 and Amd64 hosts, Go automatically generates the correct binaries.
Dockerfile Structure: All base and component Dockerfiles have been adjusted or reviewed to support multi-arch using docker. I have ensured consistency without changing the logic or layers unnecessarily.
-E2E testing - I have done E2E testing and all core harbor functionality is working.
-CI- For CI also I made plan but didnot change as docker images names need to be changes
I can make those changes and bring up a pr
Changes for CI :
Jobs that running without the matter of architecture:
auto_assign_prs.yml
codeql-analysis.yml
housekeeping-stale-issues-prs.yaml
label_check.yaml
pass-CI.yml
As the above files don’t build images or run containers - arch doesn’t matter here.
Need Multi-Arch Fixes:

Build-package.yml -
This package builds harbor component images. Currently it has only amd64 so need multi arch

CI.yml -
Brings up Harbor services and runs basic tests. Runs docker-compose up needs platform-aware images

conformance_test.yml
Runs in-depth functional + integration tests. Same compose+image pull assumptions as CI
nightly-trivy-scan.yml
Security scans Harbor images via Trivy. Needs --platform to scan both arch images
publish_release.yml
Pushes images to registry during release. Must push multi-arch manifests via buildx or manifest create

Changing the above files provide us the Github actions running on Amd64 and Arm64

Let me know if anything to be updated. Thanks !!

ranimandepudi commented:
> @ranimandepudi One general question: in phase 1, how do users consume these Arm images? In other words, what kind of installer do we plan to support for Harbor on Arm—Online, Offline, or Helm Chart?

I am not touching up the functionality so it will work for all 3 in same way so it support all of them

reasonerjt commented:
> If we release it as sandbox and people use it, I think we can address if any challenges come up ! But I can add the phase 1 and 2 plan in the proposal

For sandboxing, you can fork Harbor and release it in your forked repo.  I suggest we need to ensure the quality of the change before it's merged and released.

Schwitzd commented:
Trilled to have also arm64 image, then I can switch from Bitnami chart

chlins commented:
I suggest we divide this task into two phases. In Phase 1, we will update the relevant build scripts and introduce a new pipeline for ARM. This will allow us to run unit tests, end-to-end tests, and ensure the entire workflow operates on ARM. However, no deliverables will be shipped to Harbor users during this phase; users who wish to try it in advance can build and test it locally. Once Phase 1 is complete, we will proceed to Phase 2, where we will focus on building both ARM and AMD images and releasing them to users, including the necessary updates for docker-compose and Helm. How do you think? @ranimandepudi

wy65701436 commented:
> Hi @ranimandepudi
> 
> I had an offline discussion with @chlins — we're aligned to move forward with Harbor ARM support. I have just two general comments on the proposal:
> 
>     1. Development Phases
>        Could we clearly define the phases in the proposal as follows?
> 
> 
>     * ```
>           Phase 1: Complete code implementation and add a new CI pipeline for ARM validation.
>       ```
> 
>     * ```
>           Phase 2: Publish ARM images with a multi-arch index; ship both online and offline installers.
>       ```
> 
>     * ```
>           Phase 3: Validate the Helm chat, and publish it if necessary.
>       ```
> 
> 
>     2. Architecture Parameter
>        Please clarify that we’ll use a single parameter to specify the architecture. The CI pipeline should be able to set this parameter to build for a specific architecture or both.
>        The data flow would look like this (simplified for trivy):
>        Arch => make => compile (needed for core, jobservice, registryctl; not needed for Trivy)
>        => make build -e ARCH=**
>        => _build_trivy
>        => builder.sh
>        => Dockerfile.binary
> 
> 
> Let me know if this makes sense or needs further refinement.

@ranimandepudi Can we address those comments so we can move forward with the implementation? Please let me know if you have any concerns.

chlins commented on proposals/arm64-support-harbor.md:
```diff
@@ -0,0 +1,186 @@
+
+## References:
+
+www.arm.com/migrate
```
use `-` to split line.

ranimandepudi replied:
Done

chlins commented:
@ranimandepudi Please fix the DCO as well.

ranimandepudi commented on proposals/arm64-support-harbor.md:
```diff
@@ -0,0 +1,186 @@
+
+## RoadMap:
+
+	•	Short term (v1): Enable CI, publish multi-arch images, support both installer types.
```
No I mean medium term is for testing which we will be aligning by the next release. I will update clearly.

ranimandepudi commented on proposals/arm64-support-harbor.md:
```diff
@@ -0,0 +1,186 @@
+
+Thanks for reviewing this work! Since I'm the author of this contribution from Arm, I believe this is a high-priority topic within the Harbor community and it could be introduced as a new feature soon. A formal proposal like this document will help align on the design and technical implementation.
+
+I welcome feedback, collaboration, or questions from the Harbor community
```
DOne

ranimandepudi commented:
@chlins @OrlinVasilev @wy65701436 @OrlinVasilev I have updated the PR according to the comments. Let me know if anything to be updated !! Thanks

bupd reviewed (approved):
lgtm
